File: simpleast-core/src/test/java/com/discord/simpleast/core/ParserTest.java
package com.discord.simpleast.core;

import android.graphics.Typeface;
import android.text.style.CharacterStyle;
import android.text.style.StyleSpan;

import com.discord.simpleast.core.node.Node;
import com.discord.simpleast.core.node.StyleNode;
import com.discord.simpleast.core.node.TextNode;
import com.discord.simpleast.core.parser.Parser;
import com.discord.simpleast.core.simple.SimpleMarkdownRules;
import com.discord.simpleast.core.utils.TreeMatcher;

import junit.framework.Assert;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ParserTest {

    private Parser<Object, Node<Object>> parser;
    private TreeMatcher treeMatcher;

    @Before
    public void setup() {
        parser = new Parser<>();
        parser.addRules(SimpleMarkdownRules.createSimpleMarkdownRules());
        treeMatcher = new TreeMatcher();
        treeMatcher.registerDefaultMatchers();
    }

    @After
    public void tearDown() {
        parser = null;
    }

    @Test
    public void testEmptyParse() throws Exception {
        final List<Node<Object>> ast = parser.parse("");
        Assert.assertTrue(ast.isEmpty());
    }

    @Test
    public void testParseFormattedText() throws Exception {
        final List<Node<Object>> ast = parser.parse("**bold**");

        final StyleNode boldNode = StyleNode.Companion.createWithText("bold", Collections.singletonList((CharacterStyle) new StyleSpan(Typeface.BOLD)));

        final List<? extends Node> model = Collections.singletonList(boldNode);
        Assert.assertTrue(treeMatcher.matches(model, ast));
    }

    @Test
    public void testParseLeadingFormatting() throws Exception {
        final List<Node<Object>> ast = parser.parse("**bold** and not bold");

        final StyleNode boldNode = StyleNode.Companion.createWithText("bold", Collections.singletonList((CharacterStyle) new StyleSpan(Typeface.BOLD)));
        final TextNode trailingText = new TextNode(" and not bold");

        final List<? extends Node> model = Arrays.asList(boldNode, trailingText);
        Assert.assertTrue(treeMatcher.matches(model, ast));
    }

    @Test
    public void testParseTrailingFormatting() throws Exception {
        final List<Node<Object>> ast = parser.parse("not bold **and bold**");

        final TextNode leadingText = new TextNode("not bold ");
        final StyleNode boldNode = StyleNode.Companion.createWithText("and bold", Collections.singletonList((CharacterStyle) new StyleSpan(Typeface.BOLD)));

        final List<? extends Node> model = Arrays.asList(leadingText, boldNode);
        Assert.assertTrue(treeMatcher.matches(model, ast));
    }

    @Test
    public void testNestedFormatting() throws Exception {
//        final List<Node> ast = parser.parse("*** test1 ** test2 * test3 * test4 ** test5 ***");
        final List<Node<Object>> ast = parser.parse("**bold *and italics* and more bold**");
//        final List<Node> ast = parser.parse("______" +
//            "t"
//        + "______");

        final StyleNode<Object> boldNode = new StyleNode<>(Collections.singletonList((CharacterStyle) new StyleSpan(Typeface.BOLD)));
        boldNode.addChild(new TextNode<>("bold "));
        boldNode.addChild(StyleNode.Companion.createWithText("and italics",
            Collections.singletonList((CharacterStyle) new StyleSpan(Typeface.ITALIC))));
        boldNode.addChild(new TextNode<>(" and more bold"));

        final List<? extends Node> model = Collections.singletonList(boldNode);
        Assert.assertTrue(treeMatcher.matches(model, ast));
    }
}

File: simpleast-core/src/test/java/com/discord/simpleast/core/ParserTest.kt
package com.discord.simpleast.core

import android.graphics.Typeface
import android.text.style.CharacterStyle
import android.text.style.StyleSpan
import com.discord.simpleast.core.node.Node
import com.discord.simpleast.core.node.StyleNode
import com.discord.simpleast.core.node.TextNode
import com.discord.simpleast.core.parser.Parser
import com.discord.simpleast.core.simple.SimpleMarkdownRules
import com.discord.simpleast.core.utils.TreeMatcher
import org.junit.Assert
import org.junit.Before
import org.junit.Test


class ParserTest {
  private lateinit var parser: Parser<Any, Node<Any>, Any?>
  private lateinit var treeMatcher: TreeMatcher

  @Before
  fun setup() {
    parser = Parser<Any, Node<Any>, Any?>()
        .addRules(SimpleMarkdownRules.createSimpleMarkdownRules())
    treeMatcher = TreeMatcher()
    treeMatcher.registerDefaultMatchers()
  }

  @Test(expected = Parser.ParseException::class)
  fun testNoRuleMatch() {
    val badParser = Parser<Any, Node<Any>, Any?>()
        .addRules(SimpleMarkdownRules.createSimpleMarkdownRules(includeTextRule = false))
    badParser.parse("unmatched text", null)
  }

  @Test
  fun testEmptyParse() {
    val ast = parser.parse("", null)
    Assert.assertTrue(ast.isEmpty())
  }

  @Test
  fun testParseFormattedText() {
    val ast = parser.parse("**bold**", null)
    val model = listOf<Node<Any>>(
        StyleNode.wrapText("bold", listOf(StyleSpan(Typeface.BOLD) as CharacterStyle)),
    )
    Assert.assertTrue(treeMatcher.matches(model, ast))
  }

  @Test
  fun testParseLeadingFormatting() {
    val ast = parser.parse("**bold** and not bold", null)
    val model = listOf<Node<Any>>(
        StyleNode.wrapText("bold", listOf(StyleSpan(Typeface.BOLD) as CharacterStyle)),
        TextNode(" and not bold")
    )
    Assert.assertTrue(treeMatcher.matches(model, ast))
  }

  @Test
  fun testParseTrailingFormatting() {
    val ast = parser.parse("not bold **and bold**", null)
    val model = listOf<Node<Any>>(
        TextNode("not bold "),
        StyleNode.wrapText("and bold", listOf(StyleSpan(Typeface.BOLD) as CharacterStyle))
    )
    Assert.assertTrue(treeMatcher.matches(model, ast))
  }

  @Test
  fun testNestedFormatting() {
    val ast = parser.parse("**bold *and italics* and more bold**", null)

    val boldNode: StyleNode<Any, *> = StyleNode(listOf(StyleSpan(Typeface.BOLD) as CharacterStyle))
    boldNode.addChild(TextNode("bold "))
    boldNode.addChild(StyleNode.wrapText("and italics", listOf(StyleSpan(Typeface.ITALIC) as CharacterStyle)))
    boldNode.addChild(TextNode(" and more bold"))

    Assert.assertTrue(treeMatcher.matches(listOf(boldNode), ast))
  }

  @Test
  fun testNewlineRule() {
    val ast = parser.parse("Some text\n\n\n  \n\n\nnewline above", null)
    val model: List<Node<*>?> = listOf(
        TextNode("Some text"),
        TextNode("\n"),
        TextNode("\n"),
        TextNode<Any>("newline above"))
    Assert.assertTrue("actual $ast", treeMatcher.matches(model, ast))
  }
}